src/software/jetson_nano/services/network/network.cpp

```diff
@@ -74,3 +77,66 @@ void NetworkService::primitiveSetCallback(TbotsProto::PrimitiveSet input)
                      << "%";
     }
 }
+
+void NetworkService::logNewPrimitiveSet(TbotsProto::PrimitiveSet input)
+{
+    /* TODO: THIS TRACKS THE RECEIVED PRIMITIVES
+     *  - make sure it is in the lock
+     *  - log the new PRIMITIVE SET PROTO within a deque if it is valid (see primitive_tracker.send())
+     *  - update the timestamp on each primitive set to the current epoch time
+     */
+
+    if (primitive_set_tuple_rtt.size() >= 50)
+    {
+        LOG(WARNING) << "Too many primitive sets logged for round-trip calculations, halting log process";
+        return;
+    }
+
+    if (!primitive_set_tuple_rtt.empty()
+            && input.sequence_number() <= std::get<0>(primitive_set_tuple_rtt.back()))
+    {
+        // If the proto is older than the last received proto, then ignore it
+        return;
+    }
+    primitive_set_tuple_rtt.emplace_back(
+            std::tuple(input.sequence_number(),
+                       getCurrentEpochTimeInSeconds(),
+                       input.time_sent().epoch_timestamp_seconds())
+                       );
+}
+
+void NetworkService::updatePrimitiveSetLog(TbotsProto::RobotStatus &robot_status)
+{
+    /**
+     * TODO: Send return RobotStatus msg to Thunderscope
+     *      1) Traverse deque containing all received primitive_sets
+     *      2) If the last_handled_primitive_set for current robot_status is stored in the deque:
+     *         - delete all earlier sets
+     *         - stop & calculate the processing counter
+     *         - store the new omit_thunderloop_processing_time_sent = time_sent + processing duration
+     */
+
+    uint64_t seq_num = robot_status.last_handled_primitive_set();
+    while (seq_num <= std::get<0>(primitive_set_tuple_rtt.back()))
```

Comment:
Nit: We should be checking that the deque isn't empty first
```suggestion
    while (!primitive_set_tuple_rtt.empty() && seq_num <= std::get<0>(primitive_set_tuple_rtt.back()))
```
This is because `.back()` has underfined behaviour if there are not elements in the deque.
https://en.cppreference.com/w/cpp/container/deque/back

Reply:
Do we actually even need the check with the `back`? Doesn't `logNewPrimitiveSet` guarantee that the deque is in increasing sequence number and that the `last_handled_primitive_set` is always smaller than the largest sequence number in the deque (i.e. the element in its back)?